## Spectrometer absolute calibration for calculating the concentration of Th, K and U in rock samples

To determine the radionuclides concentrations in a given sample it's necessary to perform the espectrometer absolute calibration for the Potassium, Thorium and Uranium windows. Three calibration coefficients are calculated for each radioactive element, one in each standard.

**Useful libraries**

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### Example
**A laboratory wants to do an absolute calibration of its system and has the following standards**

We will build a table containing all the information for an absolute calibration.

In [2]:
standard = ['1(K)', '2(U)', '3(Th)']
std_table = pd.DataFrame(standard,columns=['standard'])
std_table

standard
0     1(K)
1     2(U)
2    3(Th)

In [3]:
std_table['Total mass (g)'] = [838.20, 836.53, 834.17]
std_table['K_Content (%)'] = [5, 0.5, 0.5]
std_table['U_Content (ppm)'] = [2, 15, 2]
std_table['Th_Content (ppm)'] = [5, 5, 40]

Foram feitas contagens de 30.000 s para cada padrão e para o branco onde foram obtidos os seguintes 
resultados, já subtraído o branco para as janelas:

In [4]:
std_table['Counting time (s)'] = [30000,30000,30000]
std_table['Window 1'] = [85934, 46076, 31564]
std_table['Window 2'] = [4501, 22023, 16673]
std_table['Window 3'] = [4719, 8901, 30400]

In [5]:
std_table

standard  Total mass (g)  K_Content (%)  U_Content (ppm)  Th_Content (ppm)  \
0     1(K)          838.20            5.0                2                 5   
1     2(U)          836.53            0.5               15                 5   
2    3(Th)          834.17            0.5                2                40   

   Counting time (s)  Window 1  Window 2  Window 3  
0              30000     85934      4501      4719  
1              30000     46076     22023      8901  
2              30000     31564     16673     30400

To determine the calibration coefficients, it is necessary to solve the following equations system:

$$X_j = m_{X1}.A_{jk} + m_{X2}.A_{jk} + m_{X3}.A_{jk}$$

Where j,k=[1,2,3], X=[K, U, Th] and $$A_{jk} = \frac{Window_{jk}}{Counting_{jk}}$$.

In [6]:
k = []
u = []
th = []
for i in range(len(std_table)):
    k.append((std_table['K_Content (%)'][i]/100)*std_table['Total mass (g)'][i])
    u.append((std_table['U_Content (ppm)'][i]/10e6)*std_table['Total mass (g)'][i])
    th.append((std_table['Th_Content (ppm)'][i]/10e6)*std_table['Total mass (g)'][i])

In [7]:
std_table

standard  Total mass (g)  K_Content (%)  U_Content (ppm)  Th_Content (ppm)  \
0     1(K)          838.20            5.0                2                 5   
1     2(U)          836.53            0.5               15                 5   
2    3(Th)          834.17            0.5                2                40   

   Counting time (s)  Window 1  Window 2  Window 3  
0              30000     85934      4501      4719  
1              30000     46076     22023      8901  
2              30000     31564     16673     30400

In [8]:
a1 = []
a2 = []
a3 = []
for j in range(len(std_table)):
    a1.append(std_table[f'Window {j+1}'][0]/std_table['Counting time (s)'][j])
    a2.append(std_table[f'Window {j+1}'][1]/std_table['Counting time (s)'][j])
    a3.append(std_table[f'Window {j+1}'][2]/std_table['Counting time (s)'][j])

In [9]:
A = np.array([a1, a2, a3])
A

array([[2.86446667, 0.15003333, 0.1573    ],
       [1.53586667, 0.7341    , 0.2967    ],
       [1.05213333, 0.55576667, 1.01333333]])

Once we know all the values ​​$ X_j $ we can build three **linear equations system** to be solved. The resolution of these systems presents us 9 calibration coefficients used to calculate the concentrations of radionuclides in a sample.

In [10]:
print(f'{k[0] :.2f} = mk1 . {A[0][0] :.2f} + mk2 . {A[0][1] :.2f} + mk3 . {A[0][2] :.2f}')
print(f'{k[1] :.2f} = mk1 . {A[1][0] :.2f} + mk2 . {A[1][1] :.2f} + mk3 . {A[1][2] :.2f}')
print(f'{k[2] :.2f} = mk1 . {A[2][0] :.2f} + mk2 . {A[2][1] :.2f} + mk3 . {A[2][2] :.2f}')

41.91 = mk1 . 2.86 + mk2 . 0.15 + mk3 . 0.16
4.18 = mk1 . 1.54 + mk2 . 0.73 + mk3 . 0.30
4.17 = mk1 . 1.05 + mk2 . 0.56 + mk3 . 1.01


In [11]:
print(f'{u[0] :.2f} = mu1 . {A[0][0] :.2f} + mu2 . {A[0][1] :.2f} + mu3 . {A[0][2] :.2f}')
print(f'{u[1] :.2f} = mu1 . {A[1][0] :.2f} + mu2 . {A[1][1] :.2f} + mu3 . {A[1][2] :.2f}')
print(f'{u[2] :.2f} = mu1 . {A[2][0] :.2f} + mu2 . {A[2][1] :.2f} + mu3 . {A[2][2] :.2f}')

0.00 = mu1 . 2.86 + mu2 . 0.15 + mu3 . 0.16
0.00 = mu1 . 1.54 + mu2 . 0.73 + mu3 . 0.30
0.00 = mu1 . 1.05 + mu2 . 0.56 + mu3 . 1.01


In [12]:
print(f'{th[0]} = mth1 . {A[0][0] :.2f} + mth2 . {A[0][1] :.2f} + mth3 . {A[0][2] :.2f}')
print(f'{th[1]} = mth1 . {A[1][0] :.2f} + mth2 . {A[1][1] :.2f} + mth3 . {A[1][2] :.2f}')
print(f'{th[2]} = mth1 . {A[2][0] :.2f} + mth2 . {A[2][1] :.2f} + mth3 . {A[2][2] :.2f}')

0.0004191 = mth1 . 2.86 + mth2 . 0.15 + mth3 . 0.16
0.00041826499999999997 = mth1 . 1.54 + mth2 . 0.73 + mth3 . 0.30
0.00333668 = mth1 . 1.05 + mth2 . 0.56 + mth3 . 1.01


More clearly, we have

$$(I) = \left\{\begin{matrix}
41.91 = m_{k1} . 2.86 + m_{k2} . 0.15 + m_{k3} . 0.16 \\ 
4.18 = m_{k1} . 1.54 + m_{k2} . 0.73 + m_{k3} . 0.30\\ 
4.17 = m_{k1} . 1.05 + m_{k2} . 0.56 + m_{k3} . 1.01
\end{matrix}\right.$$

$$(II) = \left\{\begin{matrix}
0.00016764 = m_{u1} . 2.86 + m_{u2} . 0.15 + m_{u3} . 0.16 \\ 
0.001254795 = m_{u1} . 1.54 + m_{u2} . 0.73 + m_{u3} . 0.30\\ 
0.0001668339 = m_{u1} . 1.05 + m_{u2} . 0.56 + m_{u3} . 1.01
\end{matrix}\right.$$

$$(III) = \left\{\begin{matrix}
0.0004191 = m_{th1} . 2.86 + m_{th2} . 0.15 + m_{th3} . 0.16 \\ 
0.0004182649 = m_{th1} . 1.54 + m_{th2} . 0.73 + m_{th3} . 0.30\\ 
0.00333668 = m_{th1} . 1.05 + m_{th2} . 0.56 + m_{th3} . 1.01
\end{matrix}\right.$$

These three equations systems give us the 9 calibration coefficients. To solve them, the **Gauss-Jacobi** iterative method was implemented.

# Gauss-Jacobi method

Consider a linear system $\textbf{A} \textbf{m}_k = \textbf{X}$, where **A** is a non-singular matrix with $a_{ii} \neq 0$ with i = [ 1, ..., n].

Isolating $m_{ki}$ from each line in the system, we write the system (**I**) as:

$$m_{k1} = \frac{(X_1 - (m_{k2} . a_{12} + m_{k3}. a_{13}))}{a_{11}} \\
m_{k2} = \frac{(X_2 - (m_{k1} . a_{21} + m_{k3}. a_{23}))}{a_{22}} \\
m_{k3} = \frac{(X_3 - (m_{k1} . a_{31} + m_{k2}. a_{32}))}{a_{33}}  $$

So we have $m_k = Cm_k + g$, where

$$
C = \begin{pmatrix}
0 & -a_{12}/a_{11} & -a_{13}/a_{11}  & \cdots  &  -a_{1n}/a_{11}\\ 
-a_{21}/a_{22} & 0 & -a_{23}/a_{22} &  \cdots & -a_{2n}/a_{22}\\ 
\vdots & \vdots & \vdots & \vdots & \vdots\\ 
-a_{n1}/a_{nn} & -a_{n2}/a_{nn} & -a_{n2}/a_{nn} & \cdots & 0 
\end{pmatrix}
$$

And

$$
g = \begin{pmatrix}
X_{1}/a_{11}\\ 
X_{2}/a_{22}\\ 
\vdots\\
X_{n}/a_{nn} 
\end{pmatrix} 
$$

Given an initial approximation $ m_{k}^{(0)}$ for the $\textbf{A}\textbf{m}_k = \textbf{x}$ system solution, the Jacobi method defines the sequence of vectors $ \left \{\textbf{x}^{(k)} \right \}_{k\geq0}$ using the following recurrence relationship:

$$m_{ki}^{K+1} = \frac{\left ( \textbf(x_i) - \sum_{j=1}^{n} a_{ij}m_{kj}^{(k)} \right )}{a_{ii}}$$

Adopting as a stopping criterion only the maximum number of iterations entered by the user, the function to solve the equation system was implemented.

In [13]:
def gaussJacobi(A, b, Niter = 40):
    aux = np.zeros(len(A))                                #auxiliary array to update the current model
    xk = np.zeros(len(A))                                 #initial model xk = [0,0,0]
    cont = 0                                          #counter to limit iterations
    for i in range(len(A)):                           #transforming the matrix Amk = x in mk = Cx + g
        for j in range(len(A)):
            if i!=j:
                A[i][j] /= A[i][i]
        b[i] /= A[i][i]
        aux[i] = b[i]
    
    while cont < Niter:                 #iteration with maximum cycles equal to Niter to update the array xk
        for i in range(len(A)):
            S = 0
            for j in range(len(A)):
                if i != j:
                    S += A[i][j] * aux[j]
            xk[i] = b[i] - S
        aux = xk[:]
        cont += 1
    return xk

**For Potassium (k)**

In [14]:
mk = gaussJacobi(A, b=k)

In [15]:
mk

array([ 15.96334881, -29.12009484,   3.51242432])

**For Uranium (U)**

In [16]:
mu = gaussJacobi(A, b=u)

In [17]:
mu

array([ 3.88969671e-05,  2.17907108e-03, -1.05461097e-03])

**For Thorium (Th)**

In [18]:
mth = gaussJacobi(A, b=th)

In [19]:
mth

array([ 0.00013609, -0.00387856,  0.00522677])

**Calculating concentrations**

To determine the concentrations we will do

$$C_k = \frac{10^2}{Mt}(m_{k1}J_1 + m_{k2}J_2 + m_{k3}J_3)$$ 

$$C_u = \frac{10^6}{Mt}(m_{u1}J_1 + m_{u2}J_2 + m_{u3}J_3)$$

$$C_{th} = \frac{10^6}{Mt}(m_{th1}J_1 + m_{th2}J_2 + m_{th3}J_3)$$



The M and t values depends on the sample from which you want to calculate the concentrations, and are, respectively, the mass (g) of the sample and the counting time (s). In addition, $ J_i, i = [1,2,3] $ are the counting windows per each second and can be calculated as

$$J = \frac{Total \ window}{Counting \ time}$$.

Características da amostra

In [20]:
def concentration(mass, t, j, mx, element):
    if element == 'k':
        cx = (100/(mass*t))*(mx[0]*(j[0]/t) + mx[1]*(j[1]/t) + mx[2]*(j[2]/t))
    else:
        cx = (1000000/(mass*t))*(mx[0]*(j[0]/t) + mx[1]*(j[1]/t) + mx[2]*(j[2]/t))
    return cx    

In [21]:
mass = 918.54
t = 7200
j = [38.287, 12.911, 17.690]
ck = concentration(mass, t, j, mk, 'k')
cu = concentration(mass, t, j, mu, 'u')
cth = concentration(mass, t, j, mth, 'th')


print(f'Concentration \t Value')
print(f'\nPotassium \t {ck} %')
print(f'\nUranium \t {cu} ppm')
print(f'\nThorium \t {cth} ppm')

Concentration 	 Value

Potassium 	 6.244687155128319e-07 %

Uranium 	 2.3031985881156043e-07 ppm

Thorium 	 9.995545304936622e-07 ppm
